### PR TITLE
Add FlashAttention 2 support

### PR DESCRIPTION
**Pull request by Jeduh**
I used [this monkeypatch](https://github.com/johnsmith0031/alpaca_lora_4bit/blob/main/monkeypatch/llama_flash_attn_monkey_patch.py) to replace LlamaAttention's forward() function globally if FLashAttention is a python module. Meaning nothings changed if FlashAttention isn't installed, and, Bloom's attention isn't affected.

Embarrassingly, from my earlier [issue](https://github.com/bigscience-workshop/petals/issues/387) those gains in performance came directly from the recent commits that drastically improved the performance. Props to the guy behind that. Now with it confirmed working, from quickly testing, The inference is slightly worse at small context and around the same for 2048 and slightly better at inference. I keeping it very vague for now because I didn't benchmark it properly but one thing I knew for sure was different was the vram usage at 2048 inference. The usage would barely move and stayed well bellow max limits whereas before, I'd be OOM'ing after a few 2048's. But now, I know it works for sure though messing up the attention mechanisms and debugging that haha. 

Could someone help me test it please? I'll see what more I can do tomorrow.

### CODE REVIEW

**Comment by borzunov** [comment]
@justheuristic @mryab Please take a look!

**Comment by mryab** [comment]
Thanks for your contribution @Jeduh! Before we review the PR, can you make sure that the style checks work? You can reformat the code by installing Black and isort and running `black . && isort .` in the root of the repository.

@borzunov Can you please point us to a set of benchmarks for throughput that would be sufficient to make sure there are no performance regressions? The benchmark scripts in the repository have no set defaults for some arguments

File: src/petals/models/llama/llama_flash_attn_monkey_patch.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+try:
+    from flash_attn.flash_attn_interface import flash_attn_func, flash_attn_varlen_qkvpacked_func
+    from flash_attn.bert_padding import unpad_input, pad_input
+except ImportError:
+    raise ImportError("Please install flash_attn to use this module")
+
+
+def replace_llama_attn_with_flash_attn():
+    transformers.models.llama.modeling_llama.LlamaAttention.forward = flash_attn_forward_gqa
+    print("Replaced attention with flash_attn")
```
Comment (mryab) [inline src/petals/models/llama/llama_flash_attn_monkey_patch.py:23]: We do not use print statements inside the library code, please use logger.debug as in other modules

Comment (Jeduh) [reply src/petals/models/llama/llama_flash_attn_monkey_patch.py]: Done, removed print and formatted with black and isort.

**Comment by borzunov** [comment]
@mryab @Jeduh I'd do benchmarks for token-by-token inference (1024 tokens in total) and for forward/backward passes with batches of 1024 tokens.

**Comment by borzunov** [comment]
Btw, does `flash-attn` work on all platforms and yields better results in all conditions? If yes - let's add this dependency to `setup.cfg` and make it a default to use Flash Attention.

**Comment by borzunov** [comment]
Re benchmarks, I actually think that the ones happening when you run a server with `--throughput eval` are enough. Please report these numbers for the old and new code

**Comment by borzunov** [comment]
Benchmarks by @Jeduh:

```
inference
 
 control
 n_steps|1_token|2      |4      |8      |16     |32     |64     |128    |256    |512    |1024   |2048   
 10      869     1382    2746    5605    10948   22596   44034   85805   180274  348232  661420  1294497 
 20      894     820     1633    3243    6385    13447   22667   39490   58863   73426   82026   81410   
 30      901     532     1063    2110    4184    8435    14795   25661   37312   45044   52432   32967   
 40      884     449     901     1783    3553    7148    12533   21610   31305   36983   43290   30546   
 50      885     413     824     1627    3226    6464    11340   19699   28107   32912   38503   27399   
 60      879     390     773     1538    3050    6082    10752   18497   26225   30212   35142   26171   
 70      865     375     746     1484    2940    5794    10213   17598   24857   28422   32721   24339   
 80      899     365     727     1438    2846    5660    9937    16858   23759   26900   30666   23005   
 90      860     357     711     1412    2784    5513    9632    16355   22702   25598   29065   23002   
 100     875     350     697     1383    2731    5386    9456    15926   21912   24508   27782   21924  
 
 flash-attn 2
 n_steps|1_token|2      |4      |8      |16     |32     |64     |128    |256    |512    |1024   |2048  
 10      829     319     639     1274    2530    4962    8896    15494   21775   25044   30074   33720   
 20      826     313     627     1241    2470    4929    8696    14922   21228   24408   29197   32444   
 30      822     308     617     1238    2440    4820    8555    14676   20795   23848   28296   29892   
 40      755     309     615     1232    2430    4823    8499    14479   20363   23212   27224   28363   
 50      787     308     612     1217    2415    4802    8366    14256   19986   22551   26172   25582   
 60      777     306     609     1217    2402    4733    8313    14113   19632   21994   25177   24646   
 70      791     305     608     1209    2396    4693    8243    13895   19091   21277   24214   23024   
 80      785     305     607     1206    2390    4670    8158    13688   18635   20572   23365   21948   


forward

 control
 n_steps|32     |64     |128    |256    |512    |1024   |2048   
 10      22819   45507   90420   188886  349520  709870  1429276 
 20      14732   23191   40459   65274   75862   103388  103466  
 30      8888    15166   26473   40029   46853   66799   72062   
 40      7380    12724   22442   33565   39127   56482   61771   
 50      6661    11775   20727   30414   35461   51176   56586   
 60      6255    11078   19573   28670   33239   48080   53135   
 70      6006    10638   18880   27520   31668   45875   50679   
 80      5874    10368   18289   26663   30672   44058   48748   
 90      5733    10226   17858   25991   29857   42538   47281   
 100     5591    9959    17491   25316   29032   41392   46050  
 
 100     5714    10097   17672   25542   29544   42579   47567   
 200     5154    9212    16143   23296   26342   37962   42878   
 300     5003    8929    15622   22252   25148   36314   41256   
 400     4888    8734    15121   21484   24309   35169   39975   
 500     4780    8548    14714   20929   23677   34329   39218   
 600     4678    8392    14426   20515   23286   33868   38618   
 700     4601    8244    14181   20152   22932   33453   38168   
 800     4548    8151    13959   19830   22560   33002   37726   
 900     4490    8059    13753   19618   22306   32597   37434   
 1000    4451    7939    13643   19450   22115   32335   37298 
 
 flash-attn 2
 n_steps|32     |64     |128    |256    |512    |1024   |2048   
 10      23812   48048   92734   193897  369313  746679  1518657 
 20      17081   26148   45277   77414   83602   101091  98045   
 30      9841    15625   27186   45250   47441   57388   60585   
 40      7973    13048   22751   37484   38680   46801   50143   
 50      7222    11855   20632   33810   35027   41777   44953   
 60      6780    11156   19490   31795   32555   38851   41753   
 70      6462    10723   18634   30416   30943   36672   39661   
 80      6283    10432   18036   29471   29882   35114   38092   
 90      6097    10174   17622   28641   28945   34043   36932   
 100     5994    9968    17305   27998   28176   33029   35895   
 
 100     6150    10253   17841   28833   29549   35586   39252   
 200     5581    9244    16114   26138   26517   31369   34645   
 300     5379    8918    15506   24995   24897   29416   32585   
 400     5245    8744    15089   24062   23923   28299   31510   
 500     5142    8548    14648   23343   23214   27480   30599   
 600     5015    8361    14286   22754   22682   26958   29955   
 700     4910    8228    13984   22265   22276   26520   29533   
 800     4843    8098    13741   21949   21941   26161   29181   
 900     4782    8002    13610   21733   21691   25851   28906   
 1000    4734    7896    13440   21508   21477   25608   28745  
```

It's really weird that we have a slowdown for more `n_steps`, gonna take a look at it that in the evening.

**Comment by justheuristic** [comment]
@borzunov gentle reminder

**Comment by Jeduh** [comment]
I identified a problem.
I have tested the same monkey patch I used here, without petals, and got significant gains in t/s in forward and inference. But the only thing I changed for this implementation was to check if the attention mask had an index size if at least 2 in the 3rd dimension before the monkey patch checked if index 2 was == 0, if i didn't, petals would've crashed as index 2 didn't exist. I realised a while ago that it's not placed there for a reason but just wanted to get it working first. The check for 0 is just a check of an autoregressive mask so I forgot about this problem towards the start because, in my head, if the attention mask was faulty then the outputs should be messed up but now its just, the attention mask is definitely not what flash-attn 2 is expecting.
Could I please ask of anyone to take a look? I still cant wrap my head around 4 dimensional attention masks, especially why are they different to begin with.

**Comment by Jeduh** [comment]
[This](https://github.com/turboderp/exllama/commit/ce687cd254cf4b1a4f266f1bf3923e8ed022e57c) is an example by another genius that I can also confirm significantly increases speeds and lowers vram usage for high token counts

**Comment by justheuristic** [comment]
Might be useful for this PR: in the latest transformers version, you can invoke flash attention like this
image

source: https://github.com/huggingface/transformers/releases/tag/v4.32.0

**Comment by Jeduh** [comment]
> Might be useful for this PR: in the latest transformers version, you can invoke flash attention like this image
> 
> source: https://github.com/huggingface/transformers/releases/tag/v4.32.0

Had my eye on BetterTransformers/Optimum for a while, finally :)
The current implementation was temporary until upstream support, now that it is, should Petals include BetterTransformers/Optimum to allow for FlashAttention2, Paged Attention and all other and future optimizations, natively? @borzunov @mryab @justheuristic

**Pull request closed without merging**